### ISSUE TITLE
Add option to complete abandoned/pending purchases

### ISSUE THREAD

**Issue by ghost1227** [easydigitaldownloads/easy-digital-downloads#2714]
Users should be able to complete abandoned/pending purchases on subsequent visits. In order to facilitate this, a status of 'processing' should be added which will be used by eChecks instead of the 'pending' status used now. This will allow pending purchases to be completed without risking affecting eCheck purchases which are pending processing.

@pippinsplugins has indicated that WooCommerce does this well and we should take inspiration from them.

**Comment by pippinsplugins** [comment]
Let's hold off on this until https://github.com/easydigitaldownloads/Easy-Digital-Downloads/issues/2667 is completed so that we are not refactoring payment creation / updates twice.

**Comment by danieliser** [comment]
Isn't this partially already covered by several extensions aimed at recovering them? Maybe would be well suited as a feature for those plugins instead. Like the idea to clean up the pending transfers though.

Also consider starting down the path with this will likely eventually result in those extensions being merged into core as first comes showing people those transactions next comes reminding them by email since that would be a simple step with this in place as well.

**Comment by pippinsplugins** [comment]
No, not really. The current extensions generate new purchase records. We want to complete the existing pending record.

**Comment by danieliser** [comment]
Oh sorry, I was under the assumption from their descriptions that they took abandoned and pending carts and notified users to come back to finish them.

**Comment by pippinsplugins** [comment]
They do notify them but they just recreate the purchase.

**Comment by danieliser** [comment]
Ok then ignore my previous statements.

**Comment by KingYes** [comment]
@pippinsplugins: For now, can I continue pending purchase manually?

**Comment by pippinsplugins** [comment]
@KingYes

**Comment by KingYes** [comment]
How?

**Comment by pippinsplugins** [comment]
Oh, sorry, I just reread what you wrote.

No, that's not supported yet.

**Comment by KingYes** [comment]
Okay.
Do you know  when it's will be supported?

**Comment by pippinsplugins** [comment]
Not yet.

**Comment by pippinsplugins** [comment]
This is too large of a feature to finish up for 2.4 since we're just a few days away from beta. Punting to 2.5.

**Comment by cklosowski** [comment]
Blocked by #2667

**Comment by cklosowski** [comment]
As much as we want this in 2.5, we're just running to close to the `release/2.5` string freeze for translation so I'm going to punt this to 2.6.

**Comment by arielk** [comment]
I'm waiting for this feature for too long :(

**Comment by pippinsplugins** [comment]
@arielk It's still very much a priority for us.

**Comment by bizzthemes** [comment]
Any news here? It's a major issue on any site we worked on that uses EDD and confuses the hell out of customers, why can't they complete the pending payment and need to create a new one. Pending payment then just sits in their account and there is nothing customers can do with it.

**Comment by pippinsplugins** [comment]
This is still on our todo list for the next major version (2.7) of EDD.

**Comment by bizzthemes** [comment]
Can we help here with the dev? It's a matter of creating a new payment with pending purchase data. To prevent double payments for 'pending' purchases, we could check API response from PayPal and if order is already marked as paid, prevent the new payment. Let me know if we're missing the bigger picture here.

**Comment by pippinsplugins** [comment]
It's a bit more complicated than that.

We'll need to:

1. Identify the pending payment ID / checkout ID in the customer's session
2. Associate the checkout ID with the payment ID
3. When submitting checkout see if the checkout ID matches an existing payment
4. If we have a payment ID, update it instead of creating a new one
5. Add a link to payment history screens to `Complete purchase` that takes customer to checkout and pre-fills checkout based on items on payment

**Comment by cklosowski** [comment]
This might be something that will truly benefit from #4196. I would say it's blocked by this actually. Since `EDD_Payment` is out of final, we can use a combination of `EDD_Payment` and an `EDD_Cart` class to easily repopulate.

**Comment by pippinsplugins** [comment]
I don't really agree.

We could very easily store an existing payment ID in session and then update it instead of creating a new one.

`EDD_Cart` is definitely a nice to have but not a requirement.

**Comment by pippinsplugins** [comment]
#2283 is required for this.

**Comment by cklosowski** [comment]
# Weekly Update
I have the cart recovery working for this. Things that are recovering.
- [ ] Single Products
- [ ] Variable Priced Products
- [ ] Item fees (like Discounts Pro)
- [ ] Global Fees
- [ ] Automatically re-selects gateway (shouldn't be a problem to change with #5082)

Still to get done here:

- Associate the purchase process to the existing payment.
- Test with Software Licensing to see if that will take a specific integration.

**Comment by pippinsplugins** [comment]
Excellent. @cklosowski Could you go ahead and open a WIP PR? I just like to watch it develop :)

**Comment by cklosowski** [comment]
In the process of doing that now actually.

Chris Klosowski
cklosowski@gmail.com
@cklosowski

On Fri, Jan 6, 2017 at 11:22 AM, Pippin Williamson <notifications@github.com

**Comment by cklosowski** [comment]
A work in progress PR is open, still needs a bit of work but should give you an idea of what it's looking like so far.

**Comment by pippinsplugins** [comment]
Sweet, looking good.

One note just in case: we need to make sure that recovering a purchase updates the existing EDD_Payment record after checkout instead of creating a new EDD_Payment.

**Comment by cklosowski** [comment]
Got that listed above in the things to do still:
`Associate the purchase process to the existing payment.`

**Comment by pippinsplugins** [comment]
Ah, missed that. :+1:

**Comment by cklosowski** [comment]
Ok, I've got this working with global cart fees, as well as fees assigned by things like discounts pro.

I'd love to have a few people test it out and see if it's working as expected. I would be sure to just use basic installs for now (no SL renewals, or upgrades, or recurring). I'm just not positive it's setup for that just yet. I'm more looking for feedback on the customer flow for this:

1) Enable PayPal Standard and Test Payment
2) Go through the checkout with PayPal, but do not complete it on the PayPal end (and do not click 'cancel and return to site')
3) Go to your purchase history on your site and find your 'pending' payment
4) Click 'Complete Purchase'
5) It should pick PayPal again by default, but with #5082 we now show the selector again, and pick test payment
6) Complete the purchase. It should be the same payment ID as from step 3.

**Comment by cklosowski** [comment]
We'll need to test out the new 'processing' status on a live site, apparently it's not possible to do eCheck in sandbox.

**Comment by pippinsplugins** [comment]
I'll try and test this tomorrow.

**Comment by cklosowski** [comment]
Seems to work just fine with SL renewals and upgrades!

**Comment by pippinsplugins** [comment]
Oh yeah! This worked great!

Could we please get a URL added to the View Order Details screen that would allow site admins to copy and paste a URL for customers to complete their purchase?

**Comment by cklosowski** [comment]
I've actually nonce'd them for a specific reason, reduces the chances that the person can replay the recovery link while logged out, it's it's generated for a logged in user. This way someone cannot start a cart recovery, go away for a couple days, then come back and trigger that link again, but logged out, thereby possibly associating the payment with a different user or something else which would mix up the data.

**Comment by pippinsplugins** [comment]
@cklosowski See comment on PR.

Noncing them makes sense but, just like happened with upgrade paths in Software Licensing, site owners are going to want to have the option of manually providing these recovery URLs. I think it's important that we support that from the get go.

**Comment by cklosowski** [comment]
I'll agree, but only if we restrict the completion of the payment to the
same user ID on the payment, if one exists.

**Comment by cklosowski** [comment]
The goal is to avoid the same issue we've been seeing where people click links for payments that should remain attached to their account but use a different email or something that triggers a weird customer experiance getting 2 customers and users created.

**Comment by pippinsplugins** [comment]
@cklosowski I agree 100%. If the pending payment has a user ID attached, let's require log in.

**Comment by cklosowski** [comment]
Ok, Things to do here:

ToDo:
- [ ] Remove Nonce from URL
- [ ] Add link to the View Order Details for admins to copy and paste
- [ ] If User ID is present on resumed payment, require login on that user before completing.

**Comment by pippinsplugins** [comment]
:+1:

**Comment by cklosowski** [comment]
@pippinsplugins ready for another round. All 4 items above have been addressed.

**Comment by pippinsplugins** [comment]
@cklosowski a few minor items to change on the PR. See review.

**Comment by cklosowski** [comment]
Ok changes made to this, my best way to do this, was to empty the cart, redirect to the checkout being empty, and set an error. Since most themes don't show an error on empty cart it doesn't give a great 'Something failed' but I feel it's better than a `wp_die` or just a random error. I can't think of many situations where someone would try and recover a payment from another account since the only place we expose it is a link in the payment history, so they'd have to copy and past it to someone else, which, I think shouldn't be expected to work.

**Comment by pippinsplugins** [comment]
@cklosowski is there any reason to not support completing a `Failed` purchase?

**Comment by pippinsplugins** [comment]
This is working really well for me.

**Comment by cklosowski** [comment]
@pippinsplugins I couldn't guarentee that `failed` hadn't already done something with the gateway. Do we know that `failed` hasn't already created records with all gateways?

**Comment by pippinsplugins** [comment]
@cklosowski the only time `Failed` created something in the gateway (that I can think of) is PayPal Express, but that's only when it fails to create or process a token. No actual payment record is created in the gateway.

**Comment by pippinsplugins** [comment]
I think `Failed` should show a `Retry Purchase` link.

**Comment by cklosowski** [comment]
@pippinsplugins sounds good, I'll add it, and keep it filterable, so if need be we can advise a gateway to remove `failed` from the list on a case by case basis.

**Comment by pippinsplugins** [comment]
:+1:

**Comment by SDavisMedia** [comment]
The UI/UX is looking and functioning wonderfully for me. I did try to complete a pending purchase through PayPal Standard, though, and everything seems to work fine until I'm sent back to my site. The payment is still pending and all the new UI for completing the purchase is still there.

<img width="739" alt="screenshot 2017-01-17 10 35 08" src="https://cloud.githubusercontent.com/assets/2359131/22034318/c2e95214-dca0-11e6-94c6-25f714a1b0a9.png">

No one else is experiencing this?

**Comment by cklosowski** [comment]
@SDavisMedia that's because PayPal standard requires that IPN pings your store to complete the purchase...and doesn't get a TXN ID until that happens. If you try and complete with a gateway that doesn't require IPN (like stripe) does it work as you expect...or test gateway?

**Comment by SDavisMedia** [comment]
Definitely works if I complete the recovery purchase with Stripe.

**Comment by cklosowski** [comment]
Thanks @SDavisMedia How we feeling here @easydigitaldownloads/core-devs, those that have tested, thinking we're in a good spot here?

**Comment by pippinsplugins** [comment]
Thought of something last night we need to support here.

1. Add items to your cart
2. Go to checkout and use PayPal Standard
3. At payPal click the back button (not cancel and return but the actual browser back button)
4. Cart will now be empty

Let's do the following:

1. Not empty the cart on submission to PayPal
2. Set the pending payment in the session to ID an in-process payment
3. Not emptying the cart means clicking back button will work and will show the existing cart contents
4. Empty the cart when we receive the `payment-processing` template upon redirect from PayPal

**Comment by SDavisMedia** [comment]
From a UI/UX perspective, that definitely couldn't hurt anything. If it's possible, I think we should definitely do it. In fact, I'm pretty sure there have been a ton of tickets about it... more so in the past.

**Comment by cklosowski** [comment]
@pippinsplugins I set it so that the cart is emptied when viewing the PayPal purchase confirmation. Clicking the back button, while it will provide the user with the cart they had before, will create a pending payment though, that isn't recoverable...

If they Go to paypal, click the back button, and then go to paypal again, they'll have 2 EDD payments, since they didn't enter the recovery process, they just clicked the back button, which in some cases, the browser never even hits the server for, just pings the local storage memory for it. So by allowing the back button, we could create more confusion, I think, as now we hve multiple orders for customers, one pending (or processing, I think is what you were tryign to say?) and one complete if they process it the 2nd time.

**Comment by pippinsplugins** [comment]
Any reason we couldn't go ahead and set it as in recovery as soon as the customer goes to PayPal? That way if they click  back, the payment ID in session is detected?

**Comment by cklosowski** [comment]
We can, however since the browser isn't actually re-polling the server or cookie data, there is a pretty high chance that it won't even re-read the cookie variable. I'll work it up to do that, however it cannot be guaranteed, so we'll have to play with it.

**Comment by cklosowski** [comment]
Ok @pippinsplugins fixed that up. It seems to work fine in Chrome at least with the tests I've done. As long as we're aware it's something that the browser 'might' not respect, I think we're ok, but it should respect it as long as the browsers are following normal specs.

**Comment by pippinsplugins** [comment]
@cklosowski just did some testing on it. Worked beautifully for me and took care of the exact problem that originally triggered us creating this issue: multiple pending payments for a single order due to customers clicking the back button.

💯

**Comment by sunnyratilal** [comment]
Tested and working perfectly fine on my end too 👍

**Comment by sunnyratilal** [comment]
Merged into `release/2.7`

**Comment by pippinsplugins** [comment]
💯

**Comment by pippinsplugins** [comment]
I'm seeing an error happen on the live site when testing this:

URL: https://easydigitaldownloads.com/checkout/?edd_action=recover_payment&payment_id=914927

Error log (incomplete for unknown reason): 

#1 /wordpress-4.7.2/wp-includes/class-wp-hook.php(323): WP_Hook->apply_filters('', Array)
#2 /wordpress-4.7.2/wp-includes/plugin.php(453): WP_Hook->do_action(Array)
#3 /dom8333/wp-content/plugins/easy-digital-downloads/includes/actions.php(25): do_action('edd_recover_pay...', Array)
#4 /wordpress-4.7.2/wp-includes/class-wp-hook.php(298): edd_get_actions('')
#5 /wordpress-4.7.2/wp-includes/class-wp-hook.php(323): WP_Hook->apply_filters(NULL, Array)
#6 /wordpress-4.7.2/wp-includes/plugin.php(453): WP_Hook->do_action(Array)
#7 /wordpress-4.7.2/wp-settings.php(449): do_action('init')
#8 /dom8333/wp-config.php(74): require_once('/wordpress-4.7....')
#9 /dom8333/wp-load.php(37): require_once('/dom8333/wp-con...')
#10 /wordpress-4.7.2/wp-blog-header.php(13): require_once('/dom8 in /dom8333/wp-content/plugins/easy-digital-downloads/includes/payments/actions.php on line 381

**Comment by pippinsplugins** [comment]
Looks like it's a left over `empty()` method call for the cart class:
```php
// Empty out the cart.
EDD()->cart->empty();
```

That method was changed to `empty_cart()`.

**Comment by pippinsplugins** [comment]
Yep, fixed.

**Comment by cklosowski** [comment]
:+1:
